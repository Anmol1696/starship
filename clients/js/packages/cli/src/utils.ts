import {
  defaultStarshipContext,
  StarshipConfig,
  StarshipContext
} from '@starship-ci/client'; // Adjust the import path as necessary
import chalk from 'chalk';
import { readFileSync } from 'fs';
import * as yaml from 'js-yaml';
import { resolve } from 'path';

import { readAndParsePackageJson } from './package';

// Function to display the version information
export function displayVersion() {
  const pkg = readAndParsePackageJson();
  console.log(chalk.green(`Name: ${pkg.name}`));
  console.log(chalk.blue(`Version: ${pkg.version}`));
}

const resolvePath = (filename: string) =>
  filename.startsWith('/') ? filename : resolve((process.cwd(), filename));

const loadYaml = (filename: string): any => {
  const path = resolvePath(filename);
  const fileContents = readFileSync(path, 'utf8');
  return yaml.load(fileContents);
};

export interface Config {
  context: StarshipContext;
  starship: StarshipConfig;
}

export const params: string[] = [
  'config',
  'name',
  'version',
  'repo',
  'repoUrl',
  'chart',
  'namespace',
  'timeout'
];

export const loadConfig = (argv: any): Config => {
  const context: StarshipContext = {
    ...defaultStarshipContext
  } as StarshipContext;
  let starship: StarshipConfig = {} as StarshipConfig;

  // Override context with command-line arguments dynamically based on StarshipContext keys
  params.forEach((key) => {
    if (argv[key] !== undefined) {
      console.log('argv:', key, ':', argv[key]);
      // @ts-expect-error - dynamic assignment
      context[key] = argv[key];
    }
  });

  if (context.config) {
    context.config = resolvePath(context.config);
    starship = loadYaml(context.config) as StarshipConfig;
  }

  return { context, starship };
};

export const usageText = `
Usage: starship <command> [options]

Commands:
  start              Start the Starship services.
  stop               Remove all components related to the deployment.
  deploy             Deploy starship using specified options or configuration file.
  setup              Setup initial configuration and dependencies.
  start-ports        Start port forwarding for the deployed services.
  stop-ports         Stop port forwarding.
  delete             Delete a specific Helm release.
  get-pods           Get the list of pods for the Helm release.
  clean              Perform a clean operation to tidy up resources.
  version, -v        Display the version of the Starship Client.

Configuration File:
  --config <path>       Specify the path to the configuration file containing the actual config file. Required.
                        Command-line options will override settings from this file if both are provided.

Command-line Options:
  --name <name>     Specify the Helm release name, default: starship.
                        Will overide config file settings for name.
  --version <ver>   Specify the version of the Helm chart, default: v0.2.6.
                        Will overide config file settings for version.
  --timeout <time>  Specify the timeout for the Helm operations, default: 10m.
                        Will overide config file settings for timeout.

Examples:
  $ starship start --config ./config/two-chain.yaml
  $ starship stop --config ./config/two-chain.yaml

If you want to setup starship for the first time
  $ starship setup

If you want to run the deployment step by step
  $ starship deploy --config ./config/two-chain.yaml
  $ starship start-ports --config ./config/two-chain.yaml
  $ starship stop-ports --config ./config/two-chain.yaml
  $ starship delete --config ./config/two-chain.yaml

Additional Help:
  $ starship help          Display this help information.
`;

export function displayUsage() {
  console.log(usageText);
}
