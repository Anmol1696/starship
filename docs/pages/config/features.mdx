# Feature Toggles

Starship allows you to have toggles for some additional
services to run with your infra setup. A few of them are
mentioned below.

## Registry

Following the schema of the `cosmos/chain-registry` repos,
if enabled this service spins up a rest api chain-registry service
for the infra you spin up.

### Syntax

```yaml
registry:
  enabled: true  # enable registry service
  localhost: true # default: true, will set chain registry output with api endpoints pointing to localhost, using the chains[].ports for localhost endpoints
  ports:
    rest: 8081   # localhost port for redirecting traffic
  # Optional: resources directive, default cpu: 0.2, memory: 200M
  resources:
    cpu: 0.5
    memory: 200M
  # Optional: image directive, default image:ghcr.io/hyperweb-io/starship/registry:20230614-7173db2
  image: ghcr.io/hyperweb-io/starship/registry:20230614-7173db2
```

> Note: `registry.localhost` is set to true by default, meaning if the ports are specified in chains[].ports, then we set the various api endpoints of the chain registry to `localhost:<port>`. Make sure `rest`, `grpc`, `rpc` are set as [chain ports](https://starship.cosmology.tech/config/chains#ports). If set to false, then it is set to kubernetes internal DNS endpoints.

### Usage

Here is a list of available endpoints and how to use them:

| Endpoint                      | Returns
| ----------------------------- |----------------------------------------------------------------------------
| `/chain_ids`                  | List of all chain-ids in the current setup
|`/chains`                      | List of all chain items
| `/chains/{chain}`             | Chain schema for the given chain ids (`name` in the `chains` directive)
| `/chains/{chain}/assets`      | Assets of the given chain
| `/chains/{chain}/keys`        | List of mnemonics used for the setup
| `/ibc`                        | List all ibc info for all chains
| `/ibc/{chain_1}/{chain_2}`    | IBC information between the 2 chains specified

Proto definition for the service is [here](https://github.com/hyperweb-io/starship/blob/main/starship/proto/registry/service.proto)

## Explorer
In order to provide a full fledged emulation environment, we have
a handy toggle to spin up an explorer for the infra.

Currently we support only Ping Pub explorer, but we will be adding more

### Syntax

```yaml
explorer:
  enabled: true  # enable registry service
  type: ping-pub # currently only support for ping-pub explorer
  ports:
    rest: 8080   # localhost port for redirecting traffic
  # Optional: resources directive, default cpu: 1, memory: 2Gi
  resources:
    cpu: 2
    memory: 4Gi
  # Optional: image directive, default image: ghcr.io/hyperweb-io/starship/ping-pub:6b7b0d096946b6bcd75d15350c7345da0d4576db
  image: ghcr.io/hyperweb-io/starship/ping-pub:6b7b0d096946b6bcd75d15350c7345da0d4576db
```

Available versions for the explorer can be found [here](https://github.com/hyperweb-io/starship/pkgs/container/starship%2Fexposer/versions?filters%5Bversion_type%5D=tagged)

### Usage

After performing `port-forward`, open explorer at: http://localhost:8080

## Ingress
In order to get external traffic into Starship, one can use the `ingress` directive to
create ingress rules on the domain.

> NOTE: Prerequists include installing `cert-issuer` and `nginx-ingress` controller in the k8s cluster.
Domain specified, needs to be pointing to the k8s cluster in which Starship is deployed
For Starship devnets, we use: [`cluster-setup.sh`](https://github.com/hyperweb-io/starship/blob/main/starship/scripts/cluster-setup.sh)
as the setup script. Need to run just once.

### Syntax
```yaml
ingress:
  enabled: true
  type: nginx
  # host must be a wildcard entry, so that we can use the wildcard to create
  # service specific ingress rules.
  host: "*.thestarship.io"
  certManager:
    issuer: "cert-issuer"
```

Above will create following endpoints with the domain, and valid certs:
* Explorer at: `https://explorere.<host>` (if enabled)
* Registry at: `https://registry.<host>` (if enabled)
* Chains at:
  * RPC endpoint: `https://rpc.<chain-id>-genesis.<host>`
  * Rest endpoint: `https://rest.<chain-id>-genesis.<host>`
  * Chain exposer: `https://rest.<chain-id>-genesis.<host>/exposer`
  * Chain faucet: `https://rest.<chain-id>-genesis.<host>/faucet`
* Relayers at:
  * Rest endpoint: `https://rest.<relayer-type>-<relayer-name>.<host>`
  * Exposer endpoint: `https://rest.<relayer-type>-<relayer-name>.<host>/exposer`

## Frontends
Starship provides flexible frontend services that can be configured to run alongside your infrastructure.
Frontend services can be customized with different images, ports, and resource allocations.

### Syntax

```yaml
frontends:
  - name: my-frontend
    type: custom
    image: ghcr.io/hyperweb-io/starship/frontend:latest
    replicas: 1
    ports:
      rest: 3000
    env:
      API_URL: http://localhost:8080
    resources:
      cpu: 0.5
      memory: 512Mi
```

### Configuration Options

| Field | Type | Description | Default |
|-------|------|-------------|---------|
| `name` | string | Name of the frontend service | Required |
| `type` | string | Type of frontend service (e.g., frontend) | Required |
| `image` | string | Docker image for the frontend service | Required |
| `replicas` | integer | Number of frontend service replicas | 1 |
| `ports.rest` | integer | Port for HTTP/REST traffic | Required |
| `env` | object | Environment variables for the frontend service | {} |
| `resources` | object | Resource limits and requests | See below |

### Resource Configuration

The `resources` field follows the standard Kubernetes resource specification:

```yaml
resources:
  limits:
    cpu: "1"
    memory: "1Gi"
  requests:
    cpu: "0.5"
    memory: "512Mi"
```

### Health Checks

The frontend service includes both readiness and liveness probes:
- TCP socket checks on the configured HTTP port
- Initial delay of 20 seconds
- Check interval of 10 seconds

### Service Configuration

The frontend service is exposed as a ClusterIP service with:
- No cluster IP (headless service)
- HTTP port matching the configured port
- Proper selector labels for service discovery

### Labels and Annotations

The frontend service includes standard Kubernetes labels and annotations:
- `app.kubernetes.io/name`: Service name
- `app.kubernetes.io/type`: Frontend type
- `app.kubernetes.io/version`: Chart version
- Quality, role, SLA, and tier annotations

### Usage

After deployment, you can access the frontend service through:
- Port forwarding: `kubectl port-forward svc/<frontend-name> <local-port>:<frontend-port>`
- Ingress (if configured): `https://<frontend-name>.<host>`

