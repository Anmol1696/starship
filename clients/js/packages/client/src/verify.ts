import axios from 'axios';

import { StarshipConfig } from './config';

export interface VerificationResult {
  service: string;
  endpoint: string;
  status: 'success' | 'failure';
  error?: string;
  details?: any;
}

export interface VerificationContext {
  config: StarshipConfig;
  localPorts: Map<string, number>;
}

export type VerificationFunction = (
  context: VerificationContext
) => Promise<VerificationResult>;

export class VerificationRegistry {
  private verifiers: Map<string, VerificationFunction[]> = new Map();

  register(service: string, verifier: VerificationFunction) {
    if (!this.verifiers.has(service)) {
      this.verifiers.set(service, []);
    }
    this.verifiers.get(service)!.push(verifier);
  }

  async run(context: VerificationContext): Promise<VerificationResult[]> {
    const results: VerificationResult[] = [];

    for (const [service, verifiers] of this.verifiers.entries()) {
      for (const verifier of verifiers) {
        try {
          const result = await verifier(context);
          results.push(result);
        } catch (error) {
          results.push({
            service,
            endpoint: 'unknown',
            status: 'failure',
            error: error instanceof Error ? error.message : 'Unknown error'
          });
        }
      }
    }

    return results;
  }
}

// Default verifiers
export const createDefaultVerifiers = (registry: VerificationRegistry) => {
  // Chain REST endpoint verification
  registry.register('chain', async (context) => {
    const results: VerificationResult[] = [];

    for (const chain of context.config.chains) {
      const port = context.localPorts.get(`${chain.id}-rest`);
      if (!port) {
        results.push({
          service: `chain-${chain.id}`,
          endpoint: 'rest',
          status: 'failure',
          error: 'Port not found'
        });
        continue;
      }

      try {
        const response = await axios.get(`http://localhost:${port}/node_info`);
        results.push({
          service: `chain-${chain.id}`,
          endpoint: 'rest',
          status: 'success',
          details: response.data
        });
      } catch (error) {
        results.push({
          service: `chain-${chain.id}`,
          endpoint: 'rest',
          status: 'failure',
          error: error instanceof Error ? error.message : 'Unknown error'
        });
      }
    }

    return results[0]; // Return first result for now
  });

  // Registry verification
  registry.register('registry', async (context) => {
    const port = context.localPorts.get('registry-rest');
    if (!port) {
      return {
        service: 'registry',
        endpoint: 'rest',
        status: 'failure',
        error: 'Port not found'
      };
    }

    try {
      const response = await axios.get(`http://localhost:${port}/status`);
      return {
        service: 'registry',
        endpoint: 'rest',
        status: 'success',
        details: response.data
      };
    } catch (error) {
      return {
        service: 'registry',
        endpoint: 'rest',
        status: 'failure',
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  });

  // Explorer verification
  registry.register('explorer', async (context) => {
    const port = context.localPorts.get('explorer-rest');
    if (!port) {
      return {
        service: 'explorer',
        endpoint: 'rest',
        status: 'failure',
        error: 'Port not found'
      };
    }

    try {
      const response = await axios.get(`http://localhost:${port}/status`);
      return {
        service: 'explorer',
        endpoint: 'rest',
        status: 'success',
        details: response.data
      };
    } catch (error) {
      return {
        service: 'explorer',
        endpoint: 'rest',
        status: 'failure',
        error: error instanceof Error ? error.message : 'Unknown error'
      };
    }
  });
};
