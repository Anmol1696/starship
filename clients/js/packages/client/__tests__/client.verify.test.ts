import axios from 'axios';

import { StarshipConfig } from '../src/config';
import { createClient, expectClient } from '../test-utils/client';
import { config } from '../test-utils/config';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('StarshipClient verify', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should verify all services successfully', async () => {
    const { client, ctx } = createClient();
    client.dependencies.forEach((dep) => (dep.installed = true));
    client.setConfig(config.config);

    // Mock successful responses for all services
    mockedAxios.get.mockImplementation((url) => {
      if (url.includes('/chains')) {
        return Promise.resolve({ status: 200, data: { chains: ['chain1'] } });
      }
      if (url.includes('/status')) {
        return Promise.resolve({ status: 200, data: { status: 'ok' } });
      }
      if (url.includes('/version')) {
        return Promise.resolve({ status: 200, data: { status: 'success' } });
      }
      if (url.includes('/config')) {
        return Promise.resolve({ status: 200, data: { chains: ['chain1'] } });
      }
      // For explorer
      return Promise.resolve({
        status: 200,
        data: '<html><body>Ping Dashboard</body></html>'
      });
    });

    await client.verify();
    expectClient(ctx, -1);
  });

  it('should handle registry verification failure', async () => {
    const { client, ctx } = createClient();
    client.dependencies.forEach((dep) => (dep.installed = true));
    client.setConfig(config.config);

    // Mock registry failure
    mockedAxios.get.mockImplementation((url) => {
      if (url.includes('/chains')) {
        return Promise.reject(new Error('Registry not available'));
      }
      // Other services succeed
      return Promise.resolve({ status: 200, data: { status: 'ok' } });
    });

    await client.verify();
    expectClient(ctx, 1);
  });

  it('should handle explorer verification failure', async () => {
    const { client, ctx } = createClient();
    client.dependencies.forEach((dep) => (dep.installed = true));
    client.setConfig(config.config);

    // Mock explorer failure
    mockedAxios.get.mockImplementation((url) => {
      if (
        !url.includes('/chains') &&
        !url.includes('/status') &&
        !url.includes('/version') &&
        !url.includes('/config')
      ) {
        return Promise.reject(new Error('Explorer not available'));
      }
      // Other services succeed
      return Promise.resolve({ status: 200, data: { status: 'ok' } });
    });

    await client.verify();
    expectClient(ctx, 1);
  });

  it('should handle relayer verification failure', async () => {
    const { client, ctx } = createClient();
    client.dependencies.forEach((dep) => (dep.installed = true));
    client.setConfig(config.config);

    // Mock relayer failure
    mockedAxios.get.mockImplementation((url) => {
      if (url.includes('/version')) {
        return Promise.reject(new Error('Relayer not available'));
      }
      // Other services succeed
      return Promise.resolve({ status: 200, data: { status: 'ok' } });
    });

    await client.verify();
    expectClient(ctx, 1);
  });

  it('should skip disabled services', async () => {
    const { client, ctx } = createClient();
    client.dependencies.forEach((dep) => (dep.installed = true));

    // Create config with disabled services
    const disabledConfig: StarshipConfig = {
      ...config.config,
      explorer: {
        ...config.config.explorer,
        enabled: false
      },
      registry: {
        ...config.config.registry,
        enabled: false
      }
    };
    client.setConfig(disabledConfig);

    // Mock successful responses for enabled services
    mockedAxios.get.mockImplementation((url) => {
      if (url.includes('/status')) {
        return Promise.resolve({ status: 200, data: { status: 'ok' } });
      }
      if (url.includes('/version')) {
        return Promise.resolve({ status: 200, data: { status: 'success' } });
      }
      return Promise.resolve({ status: 200, data: { status: 'ok' } });
    });

    await client.verify();
    expectClient(ctx, -1);
  });
});
