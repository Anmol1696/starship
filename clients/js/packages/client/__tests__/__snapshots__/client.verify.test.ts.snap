// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarshipClient verify should handle explorer verification failure 1`] = `
"Call: setConfig({"name":"osmojs","version":"v0.2.3","chains":[{"id":"osmosis-1","name":"osmosis","image":"pyramation/osmosis:v16.1.0","numValidators":1,"ports":{"rest":1317,"rpc":26657,"faucet":8007}},{"id":"cosmos-2","name":"cosmos","numValidators":1,"ports":{"rest":1313,"rpc":26653,"faucet":8003}}],"relayers":[{"name":"osmos-cosmos","type":"hermes","replicas":1,"chains":["osmosis-1","cosmos-2"]}],"registry":{"enabled":true,"ports":{"rest":8081,"grpc":9091}}})
Log: config again: [object Object]
Call: verify()
Log: Verifying services...
Log: FAILURE chain-osmosis-1 (rest): Explorer not available
Log: FAILURE chain-osmosis-1 (rpc): Block height is 0
Log: FAILURE chain-osmosis-1 (faucet): Chain faucet is not working
Log: SKIPPED chain-osmosis-1 (exposer): Port not found
Log: FAILURE chain-cosmos-2 (rest): Explorer not available
Log: FAILURE chain-cosmos-2 (rpc): Block height is 0
Log: FAILURE chain-cosmos-2 (faucet): Chain faucet is not working
Log: SKIPPED chain-cosmos-2 (exposer): Port not found
Log: SKIPPED relayer-osmos-cosmos (rest): Port not found
Log: SKIPPED relayer-osmos-cosmos (exposer): Port not found
Log: FAILURE registry (rest): Registry is not working
Log⬇

Some services failed verification. Please check the logs above.
EndLog⬆"
`;

exports[`StarshipClient verify should handle explorer verification failure 2`] = `""`;

exports[`StarshipClient verify should handle registry verification failure 1`] = `
"Call: setConfig({"name":"osmojs","version":"v0.2.3","chains":[{"id":"osmosis-1","name":"osmosis","image":"pyramation/osmosis:v16.1.0","numValidators":1,"ports":{"rest":1317,"rpc":26657,"faucet":8007}},{"id":"cosmos-2","name":"cosmos","numValidators":1,"ports":{"rest":1313,"rpc":26653,"faucet":8003}}],"relayers":[{"name":"osmos-cosmos","type":"hermes","replicas":1,"chains":["osmosis-1","cosmos-2"]}],"registry":{"enabled":true,"ports":{"rest":8081,"grpc":9091}}})
Log: config again: [object Object]
Call: verify()
Log: Verifying services...
Log: FAILURE chain-osmosis-1 (rest): Cannot read properties of undefined (reading '0')
Log: FAILURE chain-osmosis-1 (rpc): Block height is 0
Log: FAILURE chain-osmosis-1 (faucet): Chain faucet is not working
Log: SKIPPED chain-osmosis-1 (exposer): Port not found
Log: FAILURE chain-cosmos-2 (rest): Cannot read properties of undefined (reading '0')
Log: FAILURE chain-cosmos-2 (rpc): Block height is 0
Log: FAILURE chain-cosmos-2 (faucet): Chain faucet is not working
Log: SKIPPED chain-cosmos-2 (exposer): Port not found
Log: SKIPPED relayer-osmos-cosmos (rest): Port not found
Log: SKIPPED relayer-osmos-cosmos (exposer): Port not found
Log: FAILURE registry (rest): Registry not available
Log⬇

Some services failed verification. Please check the logs above.
EndLog⬆"
`;

exports[`StarshipClient verify should handle registry verification failure 2`] = `""`;

exports[`StarshipClient verify should handle relayer verification failure 1`] = `
"Call: setConfig({"name":"osmojs","version":"v0.2.3","chains":[{"id":"osmosis-1","name":"osmosis","image":"pyramation/osmosis:v16.1.0","numValidators":1,"ports":{"rest":1317,"rpc":26657,"faucet":8007}},{"id":"cosmos-2","name":"cosmos","numValidators":1,"ports":{"rest":1313,"rpc":26653,"faucet":8003}}],"relayers":[{"name":"osmos-cosmos","type":"hermes","replicas":1,"chains":["osmosis-1","cosmos-2"]}],"registry":{"enabled":true,"ports":{"rest":8081,"grpc":9091}}})
Log: config again: [object Object]
Call: verify()
Log: Verifying services...
Log: FAILURE chain-osmosis-1 (rest): Cannot read properties of undefined (reading '0')
Log: FAILURE chain-osmosis-1 (rpc): Block height is 0
Log: FAILURE chain-osmosis-1 (faucet): Chain faucet is not working
Log: SKIPPED chain-osmosis-1 (exposer): Port not found
Log: FAILURE chain-cosmos-2 (rest): Cannot read properties of undefined (reading '0')
Log: FAILURE chain-cosmos-2 (rpc): Block height is 0
Log: FAILURE chain-cosmos-2 (faucet): Chain faucet is not working
Log: SKIPPED chain-cosmos-2 (exposer): Port not found
Log: SKIPPED relayer-osmos-cosmos (rest): Port not found
Log: SKIPPED relayer-osmos-cosmos (exposer): Port not found
Log: FAILURE registry (rest): Registry is not working
Log⬇

Some services failed verification. Please check the logs above.
EndLog⬆"
`;

exports[`StarshipClient verify should handle relayer verification failure 2`] = `""`;

exports[`StarshipClient verify should skip disabled services 1`] = `
"Call: setConfig({"name":"osmojs","version":"v0.2.3","chains":[{"id":"osmosis-1","name":"osmosis","image":"pyramation/osmosis:v16.1.0","numValidators":1,"ports":{"rest":1317,"rpc":26657,"faucet":8007}},{"id":"cosmos-2","name":"cosmos","numValidators":1,"ports":{"rest":1313,"rpc":26653,"faucet":8003}}],"relayers":[{"name":"osmos-cosmos","type":"hermes","replicas":1,"chains":["osmosis-1","cosmos-2"]}],"registry":{"enabled":false,"ports":{"rest":8081,"grpc":9091}},"explorer":{"enabled":false}})
Log: config again: [object Object]
Call: verify()
Log: Verifying services...
Log: FAILURE chain-osmosis-1 (rest): Cannot read properties of undefined (reading '0')
Log: FAILURE chain-osmosis-1 (rpc): Block height is 0
Log: FAILURE chain-osmosis-1 (faucet): Chain faucet is not working
Log: SKIPPED chain-osmosis-1 (exposer): Port not found
Log: FAILURE chain-cosmos-2 (rest): Cannot read properties of undefined (reading '0')
Log: FAILURE chain-cosmos-2 (rpc): Block height is 0
Log: FAILURE chain-cosmos-2 (faucet): Chain faucet is not working
Log: SKIPPED chain-cosmos-2 (exposer): Port not found
Log: SKIPPED relayer-osmos-cosmos (rest): Port not found
Log: SKIPPED relayer-osmos-cosmos (exposer): Port not found
Log⬇

Some services failed verification. Please check the logs above.
EndLog⬆"
`;

exports[`StarshipClient verify should skip disabled services 2`] = `""`;

exports[`StarshipClient verify should verify all services successfully 1`] = `
"Call: setConfig({"name":"starship","version":"v0.2.3","chains":[{"id":"osmosis-1","name":"osmosis","image":"pyramation/osmosis:v16.1.0","numValidators":1,"ports":{"rest":1317,"rpc":26657,"faucet":8007}},{"id":"cosmos-2","name":"cosmos","numValidators":1,"ports":{"rest":1313,"rpc":26653,"faucet":8003}}],"relayers":[{"name":"osmos-cosmos","type":"hermes","replicas":1,"chains":["osmosis-1","cosmos-2"]}],"registry":{"enabled":true,"ports":{"rest":8081,"grpc":9091}}})
Log: config again: [object Object]
Call: verify()
Log: Verifying services...
Log: FAILURE chain-osmosis-1 (rest): Cannot read properties of undefined (reading '0')
Log: FAILURE chain-osmosis-1 (rpc): Block height is 0
Log: FAILURE chain-osmosis-1 (faucet): Chain faucet is not working
Log: SKIPPED chain-osmosis-1 (exposer): Port not found
Log: FAILURE chain-cosmos-2 (rest): Cannot read properties of undefined (reading '0')
Log: FAILURE chain-cosmos-2 (rpc): Block height is 0
Log: FAILURE chain-cosmos-2 (faucet): Chain faucet is not working
Log: SKIPPED chain-cosmos-2 (exposer): Port not found
Log: SKIPPED relayer-osmos-cosmos (rest): Port not found
Log: SKIPPED relayer-osmos-cosmos (exposer): Port not found
Log: SUCCESS registry (rest): Registry is working
Log⬇

Some services failed verification. Please check the logs above.
EndLog⬆"
`;

exports[`StarshipClient verify should verify all services successfully 2`] = `""`;
